libservices/CMakeLists.txt

```diff
@@ -42,3 +42,16 @@ SET(MYSQLSERVICES_SOURCES
 
 ADD_CONVENIENCE_LIBRARY(mysqlservices ${MYSQLSERVICES_SOURCES})
 INSTALL(TARGETS mysqlservices DESTINATION ${INSTALL_LIBDIR} COMPONENT Development)
+
+IF(CMAKE_SYSTEM_NAME MATCHES AIX)
+  add_custom_target(services_export
```

Comment:
this has been done before. It probably should be a `MACRO`

Reply:
It is not exactly the same.
Especially, as the exported symbols are used by plugin, in another location, file cannot be created in CMAKE_CURRENT_BINARY_DIR. However, it can be simplified as 
```
COMMAND echo "#! ."  > ${CMAKE_SOURCE_DIR}/libmysqlservices.exp
COMMAND nm -B -X32_64 -e libmysqlservices.a | awk "/^\ +[0-9]+\ [D]\ /{print\ $3}" | sed "s|\\.||g" | sort | uniq  >> ${CMAKE_SOURCE_DIR}/libmysqlservices.exp`
```

Reply:
It can be created in CMAKE_BINARY_DIR . Build should not create files in source directory, it is best to consider source directory readonly.

And btw, why are you doing all this symbol extraction?

Reply:
AIX ld is specific. All symbols must be known during link. So, when you use dlopen, you must extract symbol of dlopened libraries and put it as imported symbol where it is used.
Moreover, all symbols are not exported by default, and flag like `-Wl,-bexpall` (used to export all) can export symbols of another library (including system library).
E.g. a library **foo** is compiled with `-Wl,-bexpall` and uses `strcmp`. If a library **bar** uses **foo** and `strcmp`, it is possible `strcmp` is searched in **foo** and not in the standard library. It it not easy to predict and can have some side effect. So, I export and import symbol manually.

Reply:
Are not you repeating what cmake can already do itself?
https://cmake.org/cmake/help/v3.16/prop_tgt/ENABLE_EXPORTS.html , this is for exporting
from executable, says
"On AIX, a linker “import file” is created along with the executable to list the exported symbols for import when linking other targets. Loadable modules link to the import file to get the symbols."
exporting from executable is the only thing that is needed for plugins

Reply:
I still am a little confused about dlopen/dlsym. it is runtime "delayed" loading, from a library that is possibly not there during compilation and linking, it is different on AIX?

Reply:
> Are not you repeating what cmake can already do itself?

ENABLE_EXPORT does not export all needed symbols. It exports symbol of the executable itself, and it does not export symbols of static libraries used.

So, plugin cannot see symbols needed on mysqld, that exports only its **own** symbols. My manual export adds all symbols of all static libraries used by mysqld. Without, some plugin does not compile (I just recheck, build fails on ha_example) because they have undefined symbol (import is missing). If I add `-Wl,-berok` to bypass the error, they will crash because mysqld does not export the needed symbol by default.

> it is runtime "delayed" loading, from a library that is possibly not there during compilation and linking, it is different on AIX?

If I correctly understand your question:
- a symbol cannot be dlopened if it was not explicitly exported. ENABLE_EXPORT is not enough because symbols needed are not included in the list created by CMake.
- if load is delayed, you can use `-Wl,-berok` to ignore linking error, or import symbol. I have chosen to import symbol because it is more secure and the symbol list must be created for export, so we have it.

Reply:
this is true that ENABLE_EXPORTS does not export from static libraries. But,  you can  use
  $<TARGET_OBJECTS:sql>   $<TARGET_OBJECTS:mysys>  $<TARGET_OBJECTS:dbug> $<TARGET_OBJECTS:strings> in the source list, instead of having the libraries linked into mysqld

Reply:
As for -Wl, -berok , what were the linking errors you chose to ignore? What symbols were not  found? Maybe I will understand better if I see the errors

Reply:
For -Wl,-berok, all linking error are ignored. Example with undefined symbols and a plugin, ha_example:

```
[ 48%] Linking CXX shared module ha_example
ld: 0711-317 ERROR: Undefined symbol: PSI_server
ld: 0711-317 ERROR: Undefined symbol: .handler::ha_thd() const
   [Lot of similar message]
ld: 0711-317 ERROR: Undefined symbol: handler::set_lock_type(thr_lock_type)
ld: 0711-317 ERROR: Undefined symbol: .alloc_root
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
```

For `$<TARGET_OBJECTS:...>`, all plugin will be linked with all static library, so they will have a really larger size with this method, I think. Our goal is plugins and unit tests can see symbols of static libraries.

Reply:
Ok, these symbols should be exported from mysqld/mariadbd. Aparently, here mariadbd is also a symbol-exporting "shared object". Can you link a library to an executable, on AIX?  Can you add an option for the linker, to use  exports file for mariadbd, to make it happy?

Now, you also had berok for the test executables. What were the errors there?

Reply:
>  Can you link a library to an executable, on AIX?

The libraries are needed to create the executable. You must create the static libraries first (so, without link to the executable), and then the executable.

> Can you add an option for the linker, to use exports file for mariadbd, to make it happy?

I think it is what this line is doing:
````
TARGET_LINK_OPTIONS(mysqld PRIVATE "-Wl,-bE:${CMAKE_CURRENT_BINARY_DIR}/libmariadb.exp")
````
Just after the creation of export file.

> Now, you also had berok for the test executables. What were the errors there?

With berok, it works (add unit tests pass).
Without, 
````
[ 74%] Linking CXX executable pfs_instr-oom-t
  [ some warnings]
ld: 0711-317 ERROR: Undefined symbol: .calculate_key_len(TABLE*, unsigned int, unsigned char const*, unsigned long)
ld: 0711-317 ERROR: Undefined symbol: .compute_digest_text(sql_digest_storage const*, String*)
 [...]
ld: 0711-317 ERROR: Undefined symbol: .TABLE_SHARE::init_from_sql_statement_string(THD*, bool, char const*, unsigned long)
collect2: error: ld returned 8 exit status
gmake[2]: *** [storage/perfschema/unittest/CMakeFiles/pfs_instr-oom-t.dir/build.make:93: storage/perfschema/unittest/pfs_instr-oom-t] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:6814: storage/perfschema/unittest/CMakeFiles/pfs_instr-oom-t.dir/all] Error 2
gmake: *** [Makefile:163: all] Error 2
````
Using import file:
```
ld: 0711-407 ERROR: Symbol typeinfo name for ACL_internal_schema_access
        Visibility is not allowed on a reference to an imported symbol.
collect2: error: ld returned 8 exit status
gmake[2]: *** [storage/perfschema/unittest/CMakeFiles/pfs_instr-oom-t.dir/build.make:93: storage/perfschema/unittest/pfs_instr-oom-t] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:6839: storage/perfschema/unittest/CMakeFiles/pfs_instr-oom-t.dir/all] Error 2
gmake: *** [Makefile:163: all] Error 2
````
Note that the symbol name is "typeinfo name for ACL_internal_schema_access", I have found it in import file. We do not really understand what's happen in this case. As unit test cannot crash the server during execution (it is test, run independently), use of `-berok` is acceptable.

Reply:
The unit test will crash if you're trying to execute unit test? It can't crash the server, but if you link it with unresolved symbols, it probably won't start.. If you know you need to link with TARGET_LINK_LIBRARIES(your-test sql), link with it beter than berok

Reply:
> The unit test will crash if you're trying to execute unit test?

No. No any of them crash, most of them pass without any error.
Symbols must be exported to be used by dlopen. They are. Symbols must be imported to link without any error. For plugin, they are. For unit test, I workaround.
In this version, unit tests do not crash. In a future release, if a symbol is missing, plugins will fail during compilation, unit tests will crash during test. It is hard to be sure we have tested all plugin (so, `berok` is not acceptable), whereas unit tests are executed during test.

About TARGET_LINK_LIBRARIES, tests need symbol from different lib, not only sql. I have a lot of undefined symbols, and I do not know where they are.

Reply:
For none of the unit tests, symbols must be exported, none of the unit tests links with a shared library or say mariadbd.exe. Your unit test is just underlinked, rsp your linker fails to filter out symbols that are definitely not used during runtime.

Reply:
Why are you speaking about shared library?

>  Your unit test is just underlinked, rsp your linker fails to filter out symbols that are definitely not used during runtime.

AIX linker wants all symbols at linking time. It cannot know what is going to happen during runtime. So, symbols not used during runtime must be defined. Eg, **strings-t**, define in `unittest/strings/CMakeLists.txt`, is linked with strings and mysys by default. So, all symbols used by these two static libraries must be visible by **strings-t**. Else, compilation will fail. You can fall into a dependency hell easily with AIX linker. To avoid that, you can use `berok`, at your own risk, or import symbol (when it works).

Reply:
AIX linker wants all symbols during link time. So windows linker does. For ELF executables, it is the same. All symbols are in the static libraries strings-t is linked with (mysys, strings, and dbug, if you use debug version).
 What's the difference?

Reply:
I can imagine that some linkers can figure out that some missing external symbols from one static library  are not going to be used in resulting executable.
Maybe AIX linker is less sophisticated, but if during the linking it complains it can't find symbols, it means executable is _underlinked_. So, you could just find the static library for the missing symbol linker is complaining about. 

And the
 
   target_link_libraries(your_exe   static_library_that_defines_missing_symbol)

Reply:
I have found the missing library for unit tests. So, in `cmake/ctest.cmake`:

````
IF(CMAKE_SYSTEM_NAME MATCHES AIX)
  SET(SYMBOL_LIB_TEST sql sql_builtins)
  TARGET_LINK_LIBRARIES(${name}-t ${SYMBOL_LIB_TEST})
  ADD_DEPENDENCIES(${name}-t ${SYMBOL_LIB_TEST})
ENDIF()
````

is OK, without `berok`.